File: scripts/update-tool-layouts.py
"""
update-tool-layouts.py
Batch-updates all src/app/[lang]/tools/*/layout.tsx files to add
x-default hreflang to the alternates.languages section.
"""
import os
import re

root_dir = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
tools_dir = os.path.join(root_dir, 'src', 'app', '[lang]', 'tools')

tool_dirs = sorted([
    d for d in os.listdir(tools_dir)
    if os.path.isdir(os.path.join(tools_dir, d))
])

print(f'Found {len(tool_dirs)} tool directories to process.\n')

updated = 0
skipped = 0
errors = 0

REPLACEMENT = """languages: {{
        ...Object.fromEntries(
          i18n.locales.map((l) => [l, `https://viadreams.cc/${{l}}/tools/{tool}`])
        ),
        'x-default': `https://viadreams.cc/en/tools/{tool}`,
      }}"""

for tool_dir in tool_dirs:
    filepath = os.path.join(tools_dir, tool_dir, 'layout.tsx')
    if not os.path.exists(filepath):
        print(f'SKIP (no layout.tsx): {tool_dir}')
        skipped += 1
        continue

    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    if "'x-default'" in content or '"x-default"' in content:
        print(f'SKIP (already has x-default): {tool_dir}')
        skipped += 1
        continue

    # Extract tool name from the URL pattern
    m = re.search(r'https://viadreams\.cc/\$\{l\}/tools/([\w-]+)', content)
    if not m:
        print(f'ERROR (could not extract tool name): {tool_dir}')
        errors += 1
        continue
    tool_name = m.group(1)

    replacement = REPLACEMENT.format(tool=tool_name)

    # Pattern A: Multi-line
    pattern_a = re.compile(
        r'languages:\s*Object\.fromEntries\(\s*\n\s*i18n\.locales\.map\(\(l\)\s*=>\s*\[l,\s*`https://viadreams\.cc/\$\{l\}/tools/'
        + re.escape(tool_name)
        + r'`\]\)\s*\n\s*\)',
        re.DOTALL
    )

    # Pattern B: Single-line
    pattern_b = re.compile(
        r'languages:\s*Object\.fromEntries\(i18n\.locales\.map\(\(l\)\s*=>\s*\[l,\s*`https://viadreams\.cc/\$\{l\}/tools/'
        + re.escape(tool_name)
        + r'`\]\)\)',
        re.DOTALL
    )

    new_content = content
    matched = False

    if pattern_a.search(new_content):
        new_content = pattern_a.sub(replacement, new_content)
        matched = True
    elif pattern_b.search(new_content):
        new_content = pattern_b.sub(replacement, new_content)
        matched = True

    if matched:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(new_content)
        print(f'UPDATED: {tool_dir}')
        updated += 1
    else:
        print(f'ERROR (no pattern matched): {tool_dir}')
        errors += 1

print(f'\nDone. Updated: {updated}, Skipped: {skipped}, Errors: {errors}')

File: scripts/update-tool-layouts.js
// update-tool-layouts.js
// Batch-updates all src/app/[lang]/tools/*/layout.tsx files to add
// x-default hreflang to the alternates.languages section.

const fs = require('fs');
const path = require('path');

const rootDir = path.resolve(__dirname, '..');
const toolsDir = path.join(rootDir, 'src', 'app', '[lang]', 'tools');

const toolDirs = fs.readdirSync(toolsDir, { withFileTypes: true })
  .filter(d => d.isDirectory())
  .map(d => d.name);

console.log('Found ' + toolDirs.length + ' tool directories to process.\n');

let updatedCount = 0;
let skippedCount = 0;
let errorCount = 0;

for (const toolDir of toolDirs) {
  const filePath = path.join(toolsDir, toolDir, 'layout.tsx');

  if (!fs.existsSync(filePath)) {
    console.log('SKIP (no layout.tsx): ' + toolDir);
    skippedCount++;
    continue;
  }

  let content = fs.readFileSync(filePath, 'utf-8');

  if (content.includes("'x-default'") || content.includes('"x-default"')) {
    console.log('SKIP (already has x-default): ' + toolDir);
    skippedCount++;
    continue;
  }

  // Extract tool name from URL pattern
  const toolNameMatch = content.match(/https:\/\/viadreams\.cc\/\$\{l\}\/tools\/([\w-]+)/);
  if (!toolNameMatch) {
    console.log('ERROR (could not extract tool name): ' + toolDir);
    errorCount++;
    continue;
  }
  const toolName = toolNameMatch[1];

  let newContent = content;
  let matched = false;

  // Pattern A: Multi-line alternates block
  const multiLineRegex = new RegExp(
    'languages:\s*Object\.fromEntries\(\s*\n\s*i18n\.locales\.map\(\(l\)\s*=>\s*\[l,\s*`https://viadreams\.cc/\$\{l\}/tools/' + toolName + '`\]\)\s*\n\s*\)',
    's'
  );

  if (multiLineRegex.test(newContent)) {
    newContent = newContent.replace(multiLineRegex, function() {
      return 'languages: {\n        ...Object.fromEntries(\n          i18n.locales.map((l) => [l, `https://viadreams.cc/${l}/tools/' + toolName + '`])\n        ),\n        \'x-default\': `https://viadreams.cc/en/tools/' + toolName + '`,\n      }';
    });
    matched = true;
  }

  // Pattern B: Single-line alternates block
  if (!matched) {
    const singleLineRegex = new RegExp(
      'languages:\s*Object\.fromEntries\(i18n\.locales\.map\(\(l\)\s*=>\s*\[l,\s*`https://viadreams\.cc/\$\{l\}/tools/' + toolName + '`\]\)\)',
      's'
    );

    if (singleLineRegex.test(newContent)) {
      newContent = newContent.replace(singleLineRegex, function() {
      return 'languages: {\n        ...Object.fromEntries(\n          i18n.locales.map((l) => [l, `https://viadreams.cc/${l}/tools/' + toolName + '`])\n        ),\n        \'x-default\': `https://viadreams.cc/en/tools/' + toolName + '`,\n      }';
      });
      matched = true;
    }
  }

  if (matched) {
    fs.writeFileSync(filePath, newContent, 'utf-8');
    console.log('UPDATED: ' + toolDir);
    updatedCount++;
  } else {
    console.log('ERROR (no pattern matched): ' + toolDir);
    errorCount++;
  }
}

console.log('\nDone. Updated: ' + updatedCount + ', Skipped: ' + skippedCount + ', Errors: ' + errorCount);
